DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/GroupCoursesController.cs

```diff
@@ -0,0 +1,160 @@
+{
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Attributes;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.Models.Enums;
+    using DigitalLearningSolutions.Web.ServiceFilter;
+    using DigitalLearningSolutions.Web.ViewModels.TrackingSystem.Delegates.GroupCourses;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.FeatureManagement.Mvc;
+
+    [FeatureGate(FeatureFlags.RefactoredTrackingSystem)]
+    [Authorize(Policy = CustomPolicies.UserCentreAdmin)]
+    [SetDlsSubApplication(nameof(DlsSubApplication.TrackingSystem))]
+    [SetSelectedTab(nameof(NavMenuTab.Delegates))]
+    [Route("TrackingSystem/Delegates/Groups/{groupId:int}/Courses")]
+    public class GroupCoursesController : Controller
+    {
+        private readonly ICourseService courseService;
+        private readonly IGroupsService groupsService;
+        private readonly IUserService userService;
+
+        public GroupCoursesController(
+            IUserService userService,
+            ICourseService courseService,
+            IGroupsService groupsService
+        )
+        {
+            this.userService = userService;
+            this.courseService = courseService;
+            this.groupsService = groupsService;
+        }
+
+        [Route("{page:int=1}")]
+        public IActionResult Index(int groupId, int page = 1)
+        {
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+
+            if (groupName == null)
+            {
+                return NotFound();
+            }
+
+            var categoryIdFilter = User.GetAdminCourseCategoryFilter();
+
+            var groupCourses = groupsService.GetGroupCoursesForCategory(groupId, centreId, categoryIdFilter);
+
+            var model = new GroupCoursesViewModel(groupId, groupName, groupCourses, page);
+
+            return View(model);
+        }
+
+        [Route("{groupCustomisationId:int}/Remove")]
+        [ServiceFilter(typeof(VerifyAdminUserCanAccessGroup))]
+        [ServiceFilter(typeof(VerifyAdminUserCanAccessGroupCourse))]
+        public IActionResult RemoveGroupCourse(int groupId, int groupCustomisationId)
+        {
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+            var groupCourse = groupsService.GetUsableGroupCourseForCentre(groupCustomisationId, groupId, centreId);
+
+            var model = new RemoveGroupCourseViewModel(
+                groupCourse!.GroupCustomisationId,
+                groupCourse.CourseName,
+                groupName!
+            );
+
+            return View(model);
+        }
+
+        [HttpPost("{groupCustomisationId:int}/Remove")]
+        [ServiceFilter(typeof(VerifyAdminUserCanAccessGroup))]

```

Comment:
Can we define the `VerifyAdminUserCanAccessGroup` filter once on the controller rather than on each method?

Reply:
Yeah I think so. The Index (renamed) method that I split off didn't have it but instead has a null check within it that would do the same behaviour. I think this can be replaced with the filter which would allow us to have it on the controller.